library/ssl_tls.c

```diff
@@ -8440,6 +8444,35 @@ static int ssl_handshake_init( mbedtls_ssl_context *ssl )
     {
         ssl->handshake = mbedtls_calloc( 1, sizeof(mbedtls_ssl_handshake_params) );
     }
+#if defined(MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH)
+    /* If the buffers are too small - reallocate */
+    if( ssl->in_buf_len < MBEDTLS_SSL_IN_BUFFER_LEN )
+    {
+        resized_buffer = mbedtls_calloc( 1, MBEDTLS_SSL_IN_BUFFER_LEN );
```

Comment:
Could we use only realloc here? Realloc is better optimized because if it is possible it can use free space after current allocated memory. In that case memcpy is unnecessary and otherwise it will made memcpy himself

Reply:
In short - we might leak sensitive data on the heap if we use `realloc`, hence this approach was explicitly suggested in the optimisation design. @yanesca - is this right?